### PR TITLE
perf: optimize filterRedundantDateRanges from O(n²) to O(n log n)

### PR DESCRIPTION
**Pull request by keithwillcode**
# Refactor: Make IntervalTree Truly Generic by Moving Node Construction Outside

## Summary

This PR addresses GitHub feedback from @keithwillcode requesting that the IntervalTree be made truly generic by removing date-specific node construction from the tree class. The refactoring separates concerns by moving interval node creation logic outside the IntervalTree while maintaining identical functionality and performance.

**Key Changes:**
- **Added `createIntervalNodes` helper function** that handles date-specific node construction with start/end/maxEnd properties
- **Modified IntervalTree constructor** to accept pre-constructed `IntervalNode<T>[]` instead of raw items with extraction functions
- **Updated `filterRedundantDateRanges`** to use the new two-step approach: create nodes, then build tree
- **Preserved all existing functionality** including O(n log n) performance and ContainmentSearchAlgorithm integration

The external API of `filterRedundantDateRanges` remains unchanged, ensuring backward compatibility while enabling future extensibility for different interval types and search algorithms.

## Review & Testing Checklist for Human

- [ ] **Verify architectural improvement** - Confirm that IntervalTree is now truly generic and can be reused for non-date intervals (e.g., numeric ranges, custom interval types)
- [ ] **Test correctness with edge cases** - Run comprehensive tests with identical ranges, overlapping patterns, and the cubic-dev-ai bug scenario to ensure no regressions
- [ ] **Performance regression testing** - Benchmark the refactored code against the previous version to confirm no performance degradation from the additional `createIntervalNodes` function call
- [ ] **Validate API design** - Assess whether the separation of `createIntervalNodes` and `IntervalTree` constructor provides the intended flexibility for future interval types

**Recommended Test Plan:** Test `filterRedundantDateRanges` with enterprise-scale datasets (1000+ date ranges) and verify that both correctness and performance are maintained. Consider creating a simple non-date interval example to validate the generic design.

---

### Diagram

```mermaid
graph TD
    subgraph Legend
        L1[Major Edit]:::major-edit
        L2[Minor Edit]:::minor-edit  
        L3[Context/No Edit]:::context
    end
    
    A[filterRedundantDateRanges.ts]:::minor-edit
    B[intervalTree.ts]:::major-edit
    C[DateRange type]:::context
    D[filterRedundantDateRanges.test.ts]:::context
    
    A --> B
    A --> C
    D --> A
    
    subgraph "New Architecture"
        E[createIntervalNodes function]:::major-edit
        F[IntervalTree constructor]:::major-edit
        G[ContainmentSearchAlgorithm]:::context
    end
    
    B --> E
    B --> F
    B --> G
    A --> E
    A --> F
    
    classDef major-edit fill:#90EE90
    classDef minor-edit fill:#87CEEB  
    classDef context fill:#FFFFFF
```

### Notes

- All 12 unit tests continue to pass, including complex overlapping patterns and edge cases
- TypeScript compilation succeeds with no new errors across all 132 packages
- The refactoring maintains the existing O(n log n) worst-case complexity
- This change specifically addresses Keith's feedback about removing date-specific logic from the generic tree structure
- Future interval types (numeric ranges, time spans, etc.) can now use IntervalTree without modification

### CODE REVIEW

File: packages/lib/getAggregatedAvailability/date-range-utils/filterRedundantDateRanges.test.ts [hunk 1]
```diff
@@ -42,4 +42,73 @@ describe("filterRedundantDateRanges", () => {
...
+
+  it("should handle three identical ranges", () => {
+    const dateRanges = [
+      { start: dayjs("2025-01-23T11:00:00.000Z"), end: dayjs("2025-01-23T12:00:00.000Z") },
+      { start: dayjs("2025-01-23T11:00:00.000Z"), end: dayjs("2025-01-23T12:00:00.000Z") },
+      { start: dayjs("2025-01-23T11:00:00.000Z"), end: dayjs("2025-01-23T12:00:00.000Z") },
+    ];
+
+    const result = filterRedundantDateRanges(dateRanges);
+    expect(result.length).toBe(0);
```
Comment (keithwillcode) [inline packages/lib/getAggregatedAvailability/date-range-utils/filterRedundantDateRanges.test.ts]: DevinAI, I thought we fixed this test? This should still return 1 item of the 3 and remove the 2 redundant ones

Comment (keithwillcode) [reply packages/lib/getAggregatedAvailability/date-range-utils/filterRedundantDateRanges.test.ts]: DevinAI this test is still looking for 0 items in the list. Can you make sure this expects 1 result.

File: packages/lib/getAggregatedAvailability/date-range-utils/filterRedundantDateRanges.ts [hunk 2]
```diff
@@ -1,35 +1,111 @@
...
 
+interface IntervalNode {
+  range: DateRange;
+  index: number;
+  maxEnd: number;
+  left?: IntervalNode;
+  right?: IntervalNode;
+}
+
+class IntervalTree {
```
Comment (keithwillcode) [inline packages/lib/getAggregatedAvailability/date-range-utils/filterRedundantDateRanges.ts]: DevinAI, let’s move this and IntervalNode to a file intervalTree.ts so it can be more separated. Refactor it to use generic typing so that others can use it

File: packages/lib/getAggregatedAvailability/date-range-utils/filterRedundantDateRanges.ts [hunk 3]
```diff
@@ -1,35 +1,111 @@
...
-        otherRange.start.valueOf() <= range.start.valueOf() &&
-        otherRange.end.valueOf() >= range.end.valueOf()
+        otherRange.start.valueOf() === range.start.valueOf() &&
+        otherRange.end.valueOf() === range.end.valueOf()
       ) {
-        return false; // This range is redundant, filter it out
+        return otherIndex > index; // Keep current range only if other range has higher index
       }
+
+      return false;
```
[the remark this answers is not in the document]
Comment (keithwillcode) [reply packages/lib/getAggregatedAvailability/date-range-utils/filterRedundantDateRanges.ts:45]: DevinAI, evaluate this comment from Cubic

File: packages/lib/intervalTree.ts [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+  item: T;
+  index: number;
+  start: number;
+  end: number;
+  maxEnd: number;
+  left?: IntervalNode<T>;
+  right?: IntervalNode<T>;
+}
+
+export class IntervalTree<T> {
```
Comment (keithwillcode) [inline packages/lib/intervalTree.ts:25]: DevinAI, now that we have a generic tree and nodes, let’s use that to build the tree of nodes used for this algorithm but put the actual search algorithm in its own class so the tree doesn’t only know how to search based on dates

File: packages/lib/getAggregatedAvailability/date-range-utils/filterRedundantDateRanges.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (keithwillcode) [reply packages/lib/getAggregatedAvailability/date-range-utils/filterRedundantDateRanges.ts:45]: @cubic-dev-ai how does this look now?

File: packages/lib/getAggregatedAvailability/date-range-utils/filterRedundantDateRanges.ts [hunk 5]
```diff
@@ -1,35 +1,48 @@
...
 export function filterRedundantDateRanges(dateRanges: DateRange[]): DateRange[] {
   if (dateRanges.length <= 1) return dateRanges;
 
   const sortedRanges = [...dateRanges].sort((a, b) => a.start.valueOf() - b.start.valueOf());
+  const intervalNodes = createIntervalNodes(
+    sortedRanges,
+    (range) => range.start.valueOf(),
+    (range) => range.end.valueOf()
+  );
+  const intervalTree = new IntervalTree(intervalNodes);
```
Comment (Udit-takkar) [inline packages/lib/getAggregatedAvailability/date-range-utils/filterRedundantDateRanges.ts:19]: @keithwillcode why are sorting this twice?

here
`
const sortedRanges = [...dateRanges].sort((a, b) => a.start.valueOf() - b.start.valueOf());
`

and in packages/lib/intervalTree.ts

```
 constructor(nodes: IntervalNode<T>[]) {
    this.root = this.buildTree([...nodes].sort((a, b) => a.start - b.start));
  }
```

Comment (keithwillcode) [reply packages/lib/getAggregatedAvailability/date-range-utils/filterRedundantDateRanges.ts:19]: DevinAI, can you look at @Udit-takkar comment here and respond and/or fix it.

File: packages/lib/getAggregatedAvailability/date-range-utils/filterRedundantDateRanges.ts [hunk 6]
```diff
@@ -15,7 +15,7 @@ export function filterRedundantDateRanges(dateRanges: DateRange[]): DateRange[]
     sortedRanges,
     (range) => range.start.valueOf(),
     (range) => range.end.valueOf()
-  );
+  ).sort((a, b) => a.start - b.start);
```
Comment (keithwillcode) [inline packages/lib/getAggregatedAvailability/date-range-utils/filterRedundantDateRanges.ts]: DevinAI, is this sort still needed since we are sorting the `sortedRanges` above and then passing that into the createIntervalNodes function?

**Review by emrysal — APPROVED** [review]
LGTM, works 😎
